### PR TITLE
[4.0] Set the cli as global app

### PR DESCRIPTION
**Pull request by laoneo**
As we introduced the `CMSApplicationInterface` the cli apps should act as global applications. This is a followup of #16499 and #16516.

Additionally the apps from the controllers are loaded from the container.

~DON'T Merge, please!~

### CODE REVIEW

**Comment by laoneo** [comment]
WIll reopen when ready.

File: libraries/src/CMS/Exception/ExceptionHandler.php [hunk 1]
```diff
@@ -30,6 +30,11 @@ class ExceptionHandler
 	 */
 	public static function render($error)
 	{
+		if (\JFactory::getApplication()->isClient('cli'))
```
Comment (wilsonge) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: Should this really throw? Surely it's better to let it display json or something?

File: libraries/src/CMS/Application/CliApplication.php [hunk 2]
```diff
@@ -101,7 +101,8 @@ public function __construct(\JInputCli $input = null, Registry $config = null, C
 	 *
 	 * @return  CliApplication
 	 *
-	 * @since   11.1
+	 * @since       11.1
+	 * @deprecated  5.0 Load the app trough the container
```
Comment (wilsonge) [inline libraries/src/CMS/Application/CliApplication.php]: through

File: includes/app.php [hunk 3]
```diff
@@ -29,7 +29,7 @@
 JDEBUG ? JProfiler::getInstance('Application')->setStart($startTime, $startMem)->mark('afterLoad') : null;
 
 // Instantiate the application.
-$app = JFactory::getApplication('site');
+$app = JFactory::getContainer()->get('SiteApplication');
```
Comment (wilsonge) [inline includes/app.php]: I think this and `AdministratorApplication` still need to run through `JFactory::getApplication` otherwise we're going to end up with the `JApplicationCms::$instance` being empty right?

File: build/helpTOC.php [hunk 4]
```diff
@@ -194,11 +194,29 @@ public function doExecute()
 		$this->out('Number of strings: ' . count($toc), true);
 
 		// JSON encode the file and write it to JPATH_ADMINISTRATOR/help/en-GB/toc.json
-		file_put_contents(JPATH_ADMINISTRATOR . '/help/en-GB/toc.json', json_encode($toc));
+		file_put_contents(JPATH_ADMINISTRATOR . '/help/toc.json', json_encode($toc));
```
Comment (wilsonge) [inline build/helpTOC.php]: Why are we removing `en-GB` here?

File: libraries/src/CMS/Exception/ExceptionHandler.php (hunk 1 above)
Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: The intent with the error handler is to be Platform agnostic.  So no, this should not throw unless absolutely necessary.

File: includes/app.php (hunk 3 above)
Comment (mbabker) [inline includes/app.php]: This has to be reverted back otherwise none of the singletons will be populated.  JFactory::$application and the JApplicationCms instances container.

File: libraries/src/CMS/Service/Provider/Session.php [hunk 5]
```diff
@@ -218,7 +210,11 @@ function (Container $container)
 					}
 
 					$dispatcher = $container->get('Joomla\Event\DispatcherInterface');
-					$dispatcher->addListener('onAfterSessionStart', array(JFactory::getApplication(), 'afterSessionStart'));
+
+					if (JFactory::getApplication() instanceof CMSApplication)
```
Comment (mbabker) [inline libraries/src/CMS/Service/Provider/Session.php]: Why?  If anything this should be a `method_exists()` check.  This needlessly restricts the session service provider.

File: libraries/src/CMS/Exception/ExceptionHandler.php (hunk 1 above)
Comment (laoneo) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: What should it return then on cli?

File: build/helpTOC.php (hunk 4 above)
Comment (laoneo) [inline build/helpTOC.php]: As this folder doesn't exist, either we remove it or create it. But like it is now it crashes.

File: includes/app.php (hunk 3 above)
Comment (laoneo) [inline includes/app.php]: Can we not set them manually like I did with the cli scripts?

File: build/helpTOC.php (hunk 4 above)
Comment (mbabker) [inline build/helpTOC.php]: It's supposed to exist - https://github.com/joomla/joomla-cms/tree/staging/administrator/help/en-GB - probably something that needs to be worked on from the Mediawiki PR.

Comment (wilsonge) [inline build/helpTOC.php]: Looks like i accidently deleted it :/ oops. https://github.com/joomla/joomla-cms/commit/98cef8426a96262b2deb76425a32135d2be28d51#diff-41ac6b5cb48181a44ad59de73a6fc4dd

File: libraries/src/CMS/Exception/ExceptionHandler.php (hunk 1 above)
Comment (wilsonge) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: Or a 3rd option. Cli by default output's json rather than anything custom...

Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: Then when on CLI, it basically outputs the exception message and stack trace (trace can be optionally wrapped in a `JDEBUG` check).

I didn't set up a "fancy" exception handler for the Framework website, but basically [this logic](https://github.com/joomla/framework.joomla.org/blob/master/bin/framework#L60-L85) is what you really need (log the error and display something in the console).

File: includes/app.php (hunk 3 above)
Comment (mbabker) [inline includes/app.php]: You can set `JFactory::$application` manually, but without Reflection, you can't set `JApplicationCms::$instances`.

Comment (laoneo) [inline includes/app.php]: I will revert that change and do it then in a new pr.

File: libraries/src/CMS/Router/SiteRouter.php [hunk 6]
```diff
@@ -57,8 +58,8 @@ class SiteRouter extends Router
 	 */
 	public function __construct(CMSApplication $app = null, AbstractMenu $menu = null)
 	{
-		$this->app  = $app ?: CMSApplication::getInstance('site');
-		$this->menu = $menu ?: $this->app->getMenu();
+		$this->app  = $app ?: \JFactory::getApplication();
```
Comment (mbabker) [inline libraries/src/CMS/Router/SiteRouter.php]: I know the typehinting here is for the base `CMSApplication`, but are we certain that there isn't anything buried in the internals of the site router and menu that could cause a problem if the admin application were used in place of the site application?

File: libraries/src/CMS/Exception/ExceptionHandler.php [hunk 7]
```diff
@@ -76,7 +80,7 @@ public static function render($error)
 				 * If a type doesn't exist for that format, we try to use the format from the application's JInput object
 				 * Lastly, if all else fails, we default onto the HTML format to at least render something
 				 */
-				if (\JFactory::$document)
+				if ($isCli)
 				{
 					// We're probably in an CLI environment
```
Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: The comment here is actually very misleading, I don't think it even applies anymore.  The comment above the if statement is more accurate, so I think the change you made should be reverted.

The intent was to use the document format type if there is already a `JDocument` instance in `JFactory`, and if that isn't set to fall back on the format in the request (obviously this part is web specific).

File: libraries/src/CMS/Exception/ExceptionHandler.php [hunk 8]
```diff
@@ -98,20 +102,27 @@ public static function render($error)
...
+					// Do not allow cache
+					$app->allowCache(false);
+
+					$app->setBody($data);
 
-				echo $app->toString();
+					echo $app->toString();
 
-				$app->close(0);
+					$app->close(0);
```
Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: `$app->close()` should be called regardless of being a CLI or web app.  Or it can be removed completely as we'll hit the return line below and whatever PHP does after calling the uncaught exceptions handler would be performed.

File: libraries/src/CMS/Router/SiteRouter.php [hunk 9]
```diff
@@ -57,8 +58,8 @@ class SiteRouter extends Router
...
+		$this->menu = $menu ?: $this->app->getContainer()->get(MenuFactoryInterface::class)->createMenu('site', ['app' => $this->app]);
```
Comment (mbabker) [inline libraries/src/CMS/Router/SiteRouter.php]: Why this change?  This just seems like adding complexity for the sake of it, unless you plan on removing `getMenu()` from the `CMSApplication` class that was just fine.

Comment (laoneo) [inline libraries/src/CMS/Router/SiteRouter.php]: I'm working actually on the cli scripts. Running the indexer from the command line needs the router which needs the site menu.

File: libraries/src/CMS/Exception/ExceptionHandler.php (hunk 8 above)
Comment (laoneo) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: Then we need to add that function to the `CMSApplicationInterface`?

Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: I'm starting to think we can get away with just removing `$app->close()` completely and just using the return statement a few lines down, regardless of context.

File: libraries/src/CMS/Router/SiteRouter.php (hunk 9 above)
Comment (mbabker) [inline libraries/src/CMS/Router/SiteRouter.php]: The typehinting here is still for the web application class chain, not the interface yet.  So it's still wouldn't support a CLI app here anyway.

The next issue is the app interface isn't container aware, so you can't rely on `getContainer()` being available either.

**Comment by laoneo** [comment]
I'v changed all the input from Michael. Can you guys please have a second look?

File: libraries/src/CMS/Exception/ExceptionHandler.php [hunk 10]
```diff
@@ -98,20 +101,25 @@ public static function render($error)
...
-				$app->allowCache(false);
-
 				// If nothing was rendered, just use the message from the Exception
 				if (empty($data))
 				{
 					$data = $error->getMessage();
 				}
 
-				$app->setBody($data);
+				if ($isCli)
```
Comment (wilsonge) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: This makes me wonder if setBody and getBody should be in the interface. Thoughts @mbabker

**Comment by wilsonge** [comment]
This looks largely good. I want Michael's thoughts on the setBody/getBody thing but once we have that and we've solved conflicts I think this is probably good to merge

File: libraries/src/CMS/Exception/ExceptionHandler.php (hunk 10 above)
Comment (mbabker) [inline libraries/src/CMS/Exception/ExceptionHandler.php]: I wouldn't try finangling it to fit.  CLI and web apps handle output differently (CLI with output objects, web with response objects, both different APIs).  So this is the best option.
